cataclysm-dda: 0.G → 0.H, clean up buildInputs, fix curses build

DeeUnderscore opened the pull request:
* Add zlib. This was already present in tiles as an indirect dependency, but was missing from curses builds.
* Remove ncurses from tiles builds. This ends up in closure anyway, but CDDA itself does not require it as a direct dependency for tiles builds.
* use `--replace-fail` instead of `--replace`

Resolves #352985

## Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

DeeUnderscore committed:
cataclysm-dda: clean up buildInputs

* Add zlib. This was already present in tiles as an indirect dependency,
but was missing from curses builds.

* Remove ncurses from tiles builds. This ends up in closure anyway, but
CDDA itself does not require it as a direct dependency for tiles builds.

* use --replace-fail instead of --replace

rardiol commented:
I haven't played in a while,  but it seems the "?" keybinding menu and the Y/N confirmation prompt are broken? Is this because of the specific git version upstream is broken or something in NixOS?

`nix-build -A cataclysmDDA.git.curses`

`./result/bin/cataclysm`


![image](https://github.com/user-attachments/assets/11dc80c3-4939-40ab-ba5e-703d7661c706)
![image](https://github.com/user-attachments/assets/44b5f503-bbb8-4469-8c43-40a4b7f817da)

DeeUnderscore commented:
Yeah, this appears to be an upstream problem, see <https://github.com/CleverRaven/Cataclysm-DDA/issues/75362>. Upstream has a fix merged, but our pinned version is behind it.

rardiol commented:
Might as well update it as well then. Tested this commit, seems to work fine

```
diff --git a/pkgs/games/cataclysm-dda/git.nix b/pkgs/games/cataclysm-dda/git.nix
index 495d5f97fb09..7defdefe6631 100644
--- a/pkgs/games/cataclysm-dda/git.nix
+++ b/pkgs/games/cataclysm-dda/git.nix
@@ -2,9 +2,9 @@
 , tiles ? true, Cocoa
 , debug ? false
 , useXdgDir ? false
-, version ? "2024-07-28"
-, rev ? "bfeb1fffc4179fed242a042f24b1c97f6cfaff3d"
-, sha256 ? "sha256-IodXEA+pWfDdR9huRXieP3+J3WZJO19C8PUPT18dFBw="
+, version ? "2024-11-03"
+, rev ? "6b9889633bcd0e6a10f2f0c4c85595b48a8d5c1e"
+, sha256 ? "sha256-f7K8gYpIedMhMzEdfln/tBlLk7o0p+Orlxkk6uvEsmM="
 }:
 
 let

```

DeeUnderscore committed:
cataclysm-dda-git: 2024-07-28 -> 2024-12-11

DeeUnderscore commented:
Looks good to me, and the UI in curses seems to work fine. Thanks.

DeeUnderscore committed:
cataclysm-dda: 0.G -> 0.H

Release announcement:
https://cataclysmdda.org/releases/#0h-release-herbert

Changelog:
https://github.com/CleverRaven/Cataclysm-DDA/blob/0.H-RELEASE/data/changelog.txt

DeeUnderscore commented:
I'm adding the 0.G → 0.H bump to this PR, since it makes more sense to merge it along with the other fixes.

Dropped all the unneeded patches, and reused the same patch for Git and Stable, 

Release announcement: https://cataclysmdda.org/releases/#0h-release-herbert  
Changelog: https://github.com/CleverRaven/Cataclysm-DDA/blob/0.H-RELEASE/data/changelog.txt

rardiol reviewed (approved):
lgtm. Done a very quick test for all 4 variants

GaetanLepage commented on pkgs/games/cataclysm-dda/stable.nix:
```diff
@@ -18,49 +17,21 @@ let
       repo = "Cataclysm-DDA";
-      rev = version;
-      sha256 = "sha256-Hda0dVVHNeZ8MV5CaCbSpdOCG2iqQEEmXdh16vwIBXk=";
+      rev = "${version}-RELEASE";
```
```suggestion
      tag = "${version}-RELEASE";
```
Recently introduced.

Aleksanaa replied:
Please don't give stylistic changes with requesting change button. This is not blocking.

Aleksanaa commented:
Lemme test this a bit

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 353372`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>cataclysm-dda</li>
    <li>cataclysm-dda-git</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>cataclysm-dda</li>
    <li>cataclysm-dda-git</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>cataclysm-dda</li>
    <li>cataclysm-dda-git</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>cataclysm-dda-git</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>cataclysm-dda</li>
  </ul>
</details>

DeeUnderscore commented:
I'm bumping `-git` yet again, since with the previous commit, [upstream was failing to build for darwin as well](https://github.com/CleverRaven/Cataclysm-DDA/commit/6b9889633bcd0e6a10f2f0c4c85595b48a8d5c1e). This hopefully addresses the aarch64-darwin failure. 

As to x86_64-darwin, it's less clear to me how to solve this, seeing how it's broken in nixpkgs-unstable, and seems to be related to it taking the wrong ifdef branch and trying to use APIs not available on platform.

shazow commented:
Confirmed that this works on x86_64-linux NixOS. Seems like a strict improvement, worth merging.

FliegendeWurst commented:
Build fails (on staging-next): https://paste.fliegendewurst.eu/XYmRhb.log
```
In file included from src/third-party/flatbuffers/util.h:25,
                 from src/third-party/flatbuffers/flexbuffers.h:26,
                 from ./pch/main-pch.hpp:75,
                 from <command-line>:
src/third-party/flatbuffers/stl_emulation.h: In member function 'constexpr flatbuffers::span<T, Extent>& flatbuffers::span<T, Extent>::operator=(const flatbuffers::span<T, Extent>&)':
src/third-party/flatbuffers/stl_emulation.h:550:12: error: assignment of read-only member 'flatbuffers::span<T, Extent>::count_'
  550 |     count_ = other.count_;
      |     ~~~~~~~^~~~~~~~~~~~~~
```

DeeUnderscore commented:
* Added patch from CleverRaven/Cataclysm-DDA#74229 to stable. Built successfully against current staging-next under x86_64-linux.
* Dropped `Cocoa`/`CoreFoundation` inputs as recommended (without building on darwin)
